### PR TITLE
Triangular solver distributed (matrix A is communicated): Lower Left NoTrans case

### PR DESCRIPTION
**Pull request by Sely85**
Triangular solver distributed version, where matrix A is communicated. 
So far, only the Lower Left NoTrans case is implemented.

### CODE REVIEW

File: include/dlaf/mc/triangular_solve_distributed.h [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+            if (mat_a.rankIndex().col() == k_rank_col) {
+              auto k_local_col = distr_a.localTileFromGlobalTile<Coord::Col>(k);
+
+              auto kk = LocalTileIndex{k_local_row, k_local_col};
+
+              // Broadcast Akk row-wise
+              // Avoid useless communication if one-column communicator
+              if (row_comm_size > 1) {
+                hpx::dataflow(hpx::util::unwrapping([](auto&& tile, auto&& comm_wrapper) {
```
Comment (rasolca) [inline include/dlaf/mc/triangular_solve_distributed.h]: executors are missing in several dataflows.

File: include/dlaf/mc/triangular_solve_distributed.h [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+                                                                                      .rowCommunicator(),
+                                                                                  tile);
+                                        return std::move(tile);
+                                      }),
+                                  k_rank_col, mat_a.tileSize(GlobalTileIndex(i, k)), serial_comm());
+              }
+            }
+
+            for (SizeType j_local = 0; j_local < localnrtile_cols; ++j_local) {
+              // Matrix multiplication to update other eigenvectors
```
Comment (rasolca) [inline include/dlaf/mc/triangular_solve_distributed.h]: Eigenvector back-transformation is a specific application of the triangular solver.
In general there is no eigenvectors involved.

File: include/dlaf/solver/triangular/mc/triangular_LLN.h [hunk 3]
```diff
@@ -65,6 +65,192 @@ void triangular_LLN(blas::Diag diag, T alpha, Matrix<const T, Device::CPU>& mat_
...
+                              comm::sync::broadcast::receive_from(index,
+                                                                  comm_wrapper().colCommunicator(),
+                                                                  tile);
+                              return std::move(tile);
+                            }),
+                            k_rank_row, mat_b.tileSize(GlobalTileIndex(k, j)), serial_comm());
+        }
+      }
+
+    }  // j_local loop
```
Comment (rasolca) [inline include/dlaf/solver/triangular/mc/triangular_LLN.h]: Please avoid empty lines between braces (only exception that an empty line may be useful: to separate functions/classes closing brace from namespace closing braces).
Please also remove closing loop comments, since it is clear from the indentation what a brace closes.

File: test/unit/mc/test_triangular_solve_distributed.cpp [hunk 4]
```diff
@@ -0,0 +1,358 @@
+//
```
Comment (rasolca) [inline test/unit/mc/test_triangular_solve_distributed.cpp]: Is this file still needed?

File: test/unit/solver/mc/test_triangular.cpp [hunk 5]
```diff
@@ -26,15 +27,30 @@ using namespace dlaf::matrix::test;
...
+};
+TYPED_TEST_SUITE(TriangularSolverDistributedTest, MatrixElementTypes);
+
 std::vector<blas::Diag> blas_diags({blas::Diag::NonUnit, blas::Diag::Unit});
-std::vector<blas::Op> blas_ops({blas::Op::NoTrans, blas::Op::Trans, blas::Op::ConjTrans});
-std::vector<blas::Side> blas_sides({blas::Side::Left, blas::Side::Right});
-std::vector<blas::Uplo> blas_uplos({blas::Uplo::Lower, blas::Uplo::Upper});
+// std::vector<blas::Op> blas_ops({blas::Op::NoTrans, blas::Op::Trans, blas::Op::ConjTrans});
+// std::vector<blas::Side> blas_sides({blas::Side::Left, blas::Side::Right});
+// std::vector<blas::Uplo> blas_uplos({blas::Uplo::Lower, blas::Uplo::Upper});
```
Comment (rasolca) [inline test/unit/solver/mc/test_triangular.cpp]: In this way you also disable triangular solver local tests. (See comment below for a possible solution).

File: test/unit/solver/mc/test_triangular.cpp [hunk 6]
```diff
@@ -97,3 +156,24 @@ TYPED_TEST(TriangularSolverLocalTest, Correctness) {
...
 }
+
+TYPED_TEST(TriangularSolverDistributedTest, Correctness) {
+  SizeType m, n, mb, nb;
+
+  for (const auto& comm_grid : {this->commGrids()[0]}) {
+    for (auto diag : blas_diags) {
+      for (auto op : blas_ops) {
+        for (auto side : blas_sides) {
+          for (auto uplo : blas_uplos) {
```
Comment (rasolca) [inline test/unit/solver/mc/test_triangular.cpp]: I propose as solution ot the above comment to check here if the case is implemented:
In this case:
```
// Currently only Left Lower Notrans case is implemented
if (!(op ==NoTrans && side == Left && uplo ==Lower))
  continue;
```

Comment (rasolca) [inline test/unit/solver/mc/test_triangular.cpp]: I would reorder the loops to have the same order as in the function argument list.

File: test/unit/solver/mc/test_triangular.cpp (hunk 5 above)
Comment (Sely85) [reply test/unit/solver/mc/test_triangular.cpp]: You're right..

File: test/unit/solver/mc/test_triangular.cpp (hunk 6 above)
Comment (Sely85) [reply test/unit/solver/mc/test_triangular.cpp]: Good suggestion, thanks.
